SMLScrollView

balthisar opened the pull request:
Rather than continuing with #54, this PR takes the controversial step of adding in a custom SMLScrollView. Its scope is rather limited, but cleans up MGSFragaria quite nicely.

Namespaced SML for consistency with most of the other text view components.

Edit: Of course github is properly showing ALL of the changes from #54, too. I wonder if this #55 will clean itself up after/if you accept #54?

balthisar committed:
Exposes syntaxColouring.

balthisar committed:
Promoted the gutterView to public.

balthisar committed:
- Implemented textInvisibleCharactersColor and showsInvisibleCharacters in SMLTextView in order to remove the setter effects from MGSFragaria.
- Added some development notes/questions in the form of @todo's.

balthisar committed:
Take advantage of the new component properties to remove some hard-coded accessors.

balthisar committed:
Added a notice about KVC and properties. It may not be a bad idea to start deprecating the existing wrapper properties.

balthisar committed:
- Dynamic startingLineNumber
- Dev notes.

balthisar committed:
- Added setSyntaxDefinitionName logic to the SMLSyntaxColouring class.

balthisar committed:
Promoting to a property dynamic forwarding to the textView.

balthisar committed:
- syntaxDefinitionName now dynamic.

balthisar committed:
Sorted the dynamic declarations (just because).

balthisar committed:
Regrouped remaining accessors by component class.

balthisar committed:
Added SMLScrollView and implemented it.

balthisar committed:
scrollView accessors made dynamic.

balthisar committed:
Deprecated showsWarningsInGutter in favor of showsSyntaxErrors.

balthisar committed:
showSyntax errors should be an adjective showsSyntaxErrors.

balthisar committed:
- General comments added.

I think that the syntaxError properties are the last, genuine system-level properties (I deleted an older comment suggesting we move some of them). As such they probably do belong in MGSFragaria proper. I'm not going to touch the MGSLineNumber view, but if you would take a look at the @todo's that I added. MGSFragaria could be cleaned up a bit by getting some of the property names to match, perhaps by breaking Fragaria's API a bit and/or changing some property names or adding synonyms in MGSLineNumber view.

This is the last commit in this PR.

shysaur commented:
> this PR takes the controversial step of adding in a custom SMLScrollView. Its scope is rather limited, but cleans up MGSFragaria quite nicely.

I don't think that cleaning up _three_ properties is worth the effort of doing a subclass. The same exact result could be achieved by moving those setters in a different file, while keeping them part of the MGSFragaria class.

In the end, you moved some code, and you added some forwards to call the old code from the object where it originally belonged. In the end, the balance is that you added lines of code rather than removing them. And nothing changed either in terms of KVO compliance, so the feature balance is zero.

balthisar commented:
> By the way, sprinkling -invalidateAllColouring calls around is not necessary anymore since I switched to calling the syntax colorer in drawRect.

Okay, I added it because it was already there.

> In my opinion it's a pure waste of time to deprecate and support something that only existed for a couple days. No commit is supposed to be a release unless it is tagged as such.

and

> Again, feel free to remove/rename recently-introduced things that you want to change, instead of leaving them around and deprecating them.

Well, this will be a lot easier! I was obviously trying not to break the API for you. I agree that it's silly, but I also supposed I was being a bad citizen by breaking the functionality entirely ;-)

> In the end, you moved some code, and you added some forwards to call the old code from the object where it originally belonged. In the end, the balance is that you added lines of code rather than removing them. And nothing changed either in terms of KVO compliance, so the feature balance is zero.

Ah, yes. As I said, "controversial." I'll explain: The chief motivation wasn't cleaning up a few setters, although it had that effect. Like most of the rest of the Fragaria components the intention was to put expected behaviors in the class that's responsible for them. So the question is, is MGSFragaria really, really responsible for both `self.scrollView.hasVerticalScroller = value;` and `self.scrollView.autohidesScrollers = value;`? In the current setter, it's taking this responsibility from the scroll view. I would suggest that MGSFragaria shouldn't make this assumption, and it belongs to the scrollview, thus the subclass. _This assumes we want to keep current behavior_.

Truly the "right" thing to do would be to eliminate this setter completely and let the application developer take control of both `hasVerticalScroller` and `autohidesScrollers` instead of Fragaria making that assumption. "Legacy" behavior could be simulated in the MGSPreferencesObserver in order to keep current behavior, but this would be an API break when setting the property directly.

Given your statements above about API breaking, then I'm a lot more willing to just make the decision. However this is your repo and I'm just tagging along for the ride, so I'll probably still be a little hesitant.

(And I sincerely hope that my contributions are not more trouble than they're worth; I know from other projects the reviewing PR's takes a significant amount of time.)

shysaur commented:
Putting expected behaviors in the class that's responsible for them is a good thing.

But, in this case, it seems to me that those behaviors are already implemented in the right class (NSScrollView), because what Fragaria is doing in those methods that you moved is just adding a different means of accessing properties that already exist.

Now, if you consider the scope of NSScrollView, it's pretty generic: you can have any content view and any ruler view inside of it. Because of this, its properties' names are pretty generic, but it's OK, because it's a generic class. Even if you subclass it with a super-specific name (like `MGSPrivateFragariaOnlySpecialScrollView`) it's still an old generic scroll view, with all the genericness of its parent. It's nature didn't change. You can't even remove the old generic methods and make them private. So, it's not a good fit to have framework-specific stuff in there, because there's one thing that NSScrollView does (scrolling an NSClipView), and it does it well, and it doesn't know anything about what it scrolls (and it doesn't need to).

Instead, MGSFragaria, as a controller, knows what's in its NSScrollView and can provide some convenience methods with a name which better reflects how the various things are assembled together. Again, these methods already are a responsibility of NSScrollView, they are only given a more fitting name, which wouldn't be appropriate for NSScrollView because it's a generic class.

shysaur commented:
> So the question is, is MGSFragaria really, really responsible for both self.scrollView.hasVerticalScroller = value; and self.scrollView.autohidesScrollers = value;?

In my opinion, the answer is: MGSFragaria has never been responsible for both these two. The application can use Fragaria's convenience methods, but the scroll view has always been available to be managed manually, without any deleterious side-effects.

balthisar commented:
(Just delete long-winded response.)

Note that I'm not fighting _for_ SMLScrollView; it was just a convenience to maintain back compatibility. Instead I would prefer to break backward compatibility. The issue I see is that the expectation when seeing `hasVerticalScroller` as a convenience property is that it not also set `autohidesScroller`.

The defaults observer can simply set both properties if we want to maintain backward compatibility. 

I'm not sure if you intended to do so or not, but you've mostly convinced me that convenience properties are a waste of resources; all of the major components are exposed and directly addressable. Their only value seems to be backwards compatibility.

shysaur commented:
> Note that I'm not fighting for SMLScrollView

I know: you just told me the reasons why you introduced it, and I told you the reasons why I didn't want to introduce it.

> I'm not sure if you intended to do so or not, but you've mostly convinced me that convenience properties are a waste of resources

I'm not so extreme; convenience properties are useful (look at NSTextView: a lot of its methods are just conveniences for stuff that is implemented in NSLayoutManager, NSTextStorage, NSUndoManager, and even NSScrollView).

But finding the right balance between useful and superfluous is hard: for example, setting the colors for syntax highlighting is something you do rarely, so having a convenience method doesn't add much. For the other methods, I'm not sure yet if they are worth of being there or not, given that they are just forwards.

> The issue I see is that the expectation when seeing hasVerticalScroller as a convenience property is that it not also set autohidesScroller.

So, let's screw backcompat for such an insignificant thing: I'll remove setting autohidesScroller in that method. (actually I already wanted to do that someday)

shysaur commented:
> I'll remove setting autohidesScroller 

Done in e365e68.

balthisar commented:
Thanks. Dust-binning this PR.
